### PR TITLE
Bug/634/auto fit map after unfocussing map

### LINKED ISSUE #634: Bug auto-fit map after unfocussing map
# Bug

## Expected Behavior
**When** *showing complete map*
**after** focussing some nodes
**then** map should be auto-fitted

## Actual Behavior
**When** *showing complete map*
**after** focussing some nodes
**then** map is not auto-fitted

## Steps to Reproduce the Problem
  1. click *Focus* in nodeContextMenu on a random node
  1. click *show complete map* button on the bottom right
  1. now the map does not auto fit correctly

## Specifications

  - Version: 1.28

### PR DESCRIPTION
**Pull request by JaiJoPa**
# Auto fit map after unfocussing map

closes #634

## Description

When a package/building is focussed and then unfocussed, the camera auto fits to the map.

## Definition of Done

A task/pull request will not be considered to be complete until all these items can be checked off.

- [x] **All** requirements mentioned in the issue are implemented
- [x] Does match the Code of Conduct and the Contribution file
- [x] Task has its own **GitHub issue** (something it is solving)
  - [x] Issue number is **included in the commit messages** for traceability
- [x] **Update the README.md** with any changes/additions made
- [x] **Update the CHANGELOG.md** with any changes/additions made
- [x] **Enough test coverage to ensure that uncovered changes do not break functionality**
- [x] **All tests pass**
- [x] **Descriptive pull request text**, answering:
  - What problem/issue are you fixing?
  - What does this PR implement and how?
- [x] **Assign your PR to someone** for a code review
  - This person _will be contacted **first**_ if a bug is introduced into `master`
- [ ] **Manual testing** did not fail

### CODE REVIEW

File: visualization/app/codeCharta/state/settingsService/settings.service.events.ts [hunk 1]
```diff
@@ -32,6 +32,14 @@ export interface SearchPatternSubscriber {
 	onSearchPatternChanged(searchPattern: string)
 }
 
+export interface FocusNodePathSubscriber {
```
Comment (NearW) [inline visualization/app/codeCharta/state/settingsService/settings.service.events.ts]: Let's rename this to `FocusNodeSubscriber` and `UnfocusNodeSubscriber`
and therefore `onFocusNodePath`  to `onFocusNode`
and `NODE_PATH_FOCUSED` to `NODE_FOCUSED_EVENT`

File: visualization/app/codeCharta/state/settingsService/settings.service.events.ts [hunk 2]
```diff
@@ -32,6 +32,14 @@ export interface SearchPatternSubscriber {
...
+	onFocusNodePath(focusedNodePath: string)
+}
+
+export interface UnfocusNodePathSubscriber {
+	onUnfocusNodePath(unfocusNodePath: string)
```
Comment (NearW) [inline visualization/app/codeCharta/state/settingsService/settings.service.events.ts]: an unfocusedNode has no path. Just call it `focusedNodePath`
We could also remove that parameter from this method completly, because we know that whenever this is called, the focusedNodePath will be empty.

Comment (alschmut) [reply visualization/app/codeCharta/state/settingsService/settings.service.events.ts]: Yes, I would remove the parameter as well.

File: visualization/app/codeCharta/ui/codeMap/threeViewer/threeOrbitControlsService.ts [hunk 3]
```diff
@@ -12,29 +14,74 @@ export interface CameraChangeSubscriber {
 /**
  * Service to manage the three orbit controls in an angular way.
  */
-export class ThreeOrbitControlsService {
+export class ThreeOrbitControlsService implements FocusNodeSubscriber, UnfocusNodeSubscriber {
 	public static CAMERA_CHANGED_EVENT_NAME = "camera-changed"
 
 	public controls: OrbitControls
 	public defaultCameraPosition: Vector3 = new Vector3(0, 0, 0)
+	public defaultZoom: number = 0
```
Comment (alexhunziker) [inline visualization/app/codeCharta/ui/codeMap/threeViewer/threeOrbitControlsService.ts]: does this have to be public?

File: visualization/app/codeCharta/ui/codeMap/threeViewer/threeOrbitControlsService.ts [hunk 4]
```diff
@@ -56,30 +103,52 @@ export class ThreeOrbitControlsService {
...
-		this.defaultCameraPosition = cameraReference.clone().position
+		this.threeCameraService.camera.lookAt(t)
 
 		this.threeCameraService.camera.updateProjectionMatrix()
 	}
 
+	private initializeDefaultZoomWithoutAutoFit() {
+		const autoFittedPerspective: Vector3 = new Vector3(0, 0, 0)
+		const boundingSphere = this.getBoundingSphere()
+		const len: number = this.autoFitToCalculation(boundingSphere)
```
Comment (alexhunziker) [inline visualization/app/codeCharta/ui/codeMap/threeViewer/threeOrbitControlsService.ts]: What the function does is to calculate some sort of length, maybe use a descriptive name of what is actually calculated here as method name instead of AutoFitToCalculation?

File: visualization/app/codeCharta/ui/codeMap/threeViewer/threeOrbitControlsService.ts [hunk 5]
```diff
@@ -56,30 +103,52 @@ export class ThreeOrbitControlsService {
...
+
+		autoFittedPerspective.set(len, len, len)
+
+		this.defaultCameraPosition = autoFittedPerspective.clone()
+
+		const t: Vector3 = boundingSphere.center
```
Comment (alexhunziker) [inline visualization/app/codeCharta/ui/codeMap/threeViewer/threeOrbitControlsService.ts]: What is t?

File: visualization/app/codeCharta/ui/codeMap/threeViewer/threeOrbitControlsService.ts [hunk 6]
```diff
@@ -56,30 +103,52 @@ export class ThreeOrbitControlsService {
...
+		t.setY(0)
+
+		const centerVector: Vector3 = new Vector3(t.x, t.y, t.z)
```
Comment (alexhunziker) [inline visualization/app/codeCharta/ui/codeMap/threeViewer/threeOrbitControlsService.ts]: why do we need a new Vector here? why not just assign t to this?

File: visualization/app/codeCharta/ui/codeMap/threeViewer/threeOrbitControlsService.ts [hunk 7]
```diff
@@ -56,30 +103,52 @@ export class ThreeOrbitControlsService {
...
+
+		this.defaultZoom = autoFittedPerspective.distanceTo(centerVector)
+	}
+
+	private getBoundingSphere() {
```
Comment (alexhunziker) [inline visualization/app/codeCharta/ui/codeMap/threeViewer/threeOrbitControlsService.ts]: Could it be that the bounding box here is not correct for focused nodes? (The new issue)

File: visualization/app/codeCharta/ui/codeMap/threeViewer/threeOrbitControlsService.spec.ts [hunk 8]
```diff
@@ -91,32 +100,59 @@ describe("ThreeOrbitControlsService", () => {
...
+
+	describe("autoFitTo", () => {
+		it("should auto fit map to its origin value ", () => {
+			threeCameraService.camera.position.set(0, 0, 0)
 
-		expect(threeCameraService.camera.position.set).toBeCalledWith(vector.x, vector.y, vector.z)
-		expect(threeOrbitControlsService.controls.update).toBeCalled()
-		expect(threeCameraService.camera.lookAt).toBeCalledWith(new THREE.Vector3(0, 0, 0))
-		expect(threeCameraService.camera.updateProjectionMatrix).toBeCalled()
-		expect(threeOrbitControlsService.defaultCameraPosition.x).toEqual(vector.x)
```
Comment (alexhunziker) [inline visualization/app/codeCharta/ui/codeMap/threeViewer/threeOrbitControlsService.spec.ts]: Why did you remove these?

File: visualization/app/codeCharta/ui/loadingGif/loadingGif.service.ts [hunk 9]
```diff
@@ -25,6 +25,14 @@ export class LoadingGifService {
 		this.notifyLoadingMapFlagChange()
 	}
 
+	public isLoadingNewFile(): boolean {
```
Comment (alexhunziker) [inline visualization/app/codeCharta/ui/loadingGif/loadingGif.service.ts]: as mentioned above, i am not sure if this belongs here conceptually. The UI component responsible for showing some sort of loading indicator should not be responsible for felling other components if we are loading a map. Don't we have another way of knowing if a map is being loaded?

File: visualization/app/codeCharta/ui/codeMap/threeViewer/threeOrbitControlsService.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (JaiJoPa) [reply visualization/app/codeCharta/ui/codeMap/threeViewer/threeOrbitControlsService.ts]: No, you are right. I changed it to private :)

File: visualization/app/codeCharta/ui/codeMap/threeViewer/threeOrbitControlsService.ts [hunk 10]
```diff
@@ -12,29 +14,74 @@ export interface CameraChangeSubscriber {
...
+		SettingsService.subscribeToFocusNode($rootScope, this)
+		SettingsService.subscribeToUnfocusNode($rootScope, this)
+	}
+
+	public onFocusNode(focusPath: string) {
+		this.autoFitTo()
+	}
+
+	public onUnfocusNode() {
+		if (!this.loadingGifService.isLoadingNewFile() && !this.loadingGifService.isLoadingNewMap()) {
```
[the remark this answers is not in the document]
Comment (JaiJoPa) [reply visualization/app/codeCharta/ui/codeMap/threeViewer/threeOrbitControlsService.ts]: You are totally right. It should have a different name. I renamed it, and moved it into the state service folder.

File: visualization/app/codeCharta/ui/codeMap/threeViewer/threeOrbitControlsService.ts (hunk 4 above)
Comment (JaiJoPa) [reply visualization/app/codeCharta/ui/codeMap/threeViewer/threeOrbitControlsService.ts]: Changed it to a more descriptive name :)

File: visualization/app/codeCharta/ui/codeMap/threeViewer/threeOrbitControlsService.ts (hunk 6 above)
Comment (JaiJoPa) [reply visualization/app/codeCharta/ui/codeMap/threeViewer/threeOrbitControlsService.ts]: I don't know, why I assigned it to another vector... :/ I removed it.

File: visualization/app/codeCharta/ui/codeMap/threeViewer/threeOrbitControlsService.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (alschmut) [reply visualization/app/codeCharta/ui/codeMap/threeViewer/threeOrbitControlsService.ts]: I renamed it again to follow our convention from `loadingStatusService.ts` to `loadingStatus.service.ts`

**Comment by alschmut** [comment]
@fatihi do you want to take another look at this branch and its current focussing behaviour? If I remember right, you requested to not reset the complete camera, when unfocussing, but only the zoom. Does this look like you expected it?

In my opinion, I don't like the solution too much. It rather confuses me, as we do not have a zoom-animation, but just jump to the new view. Maybe this were an aspect to investigate: 
- Having a short animation when zooming in and out. This gives the user the possibility to track the wanted nodes.

**Comment by alschmut** [comment]
After talking to @fatihi and @JaiJoPa we decided to undo new unfocus-functionality, but keep the bug fix and other refactorings. Together with a new issue we will research in a new focus/autofit functionality with an optional camera animation.
